internal/monitors/jaegergrpc/monitor.go

```diff
@@ -0,0 +1,168 @@
+	CertFile string `yaml:"certFile"`
+	// The key file to use for tls
+	KeyFile string `yaml:"keyFile"`
+}
+
+func (tls *TLSCreds) Credentials() (credentials.TransportCredentials, error) {
+	var creds credentials.TransportCredentials
+	var err error
+	if tls != nil {
+		creds, err = credentials.NewServerTLSFromFile(tls.CertFile, tls.KeyFile)
+	}
+
+	return creds, err
+}
+
+// Config for this monitor
+type Config struct {
+	config.MonitorConfig `yaml:",inline" acceptsEndpoints:"false" singleInstance:"true"`
+	// The host:port on which to listen for traces.
+	ListenAddress string `yaml:"listenAddress" default:"0.0.0.0:14250"`
+	// TLS are optional tls credential settings to configure the GRPC server with
+	TLS *TLSCreds `yaml:"tls,omitempty"`
+}
+
+// Monitor that accepts and forwards SignalFx data
+type Monitor struct {
+	Output types.Output
+	grpc   *grpc.Server
+	ln     net.Listener
+}
+
+var _ api_v2.CollectorServiceServer = (*Monitor)(nil)
+
+func extractRemoteAddressToContext(ctx context.Context) (net.IP, bool) {
+	var sourceIP net.IP
+
+	// get peer connection info from grpc context
+	p, hasSource := peer.FromContext(ctx)
+	if !hasSource || p.Addr.String() == "" {
+		return nil, false
+	}
+
+	// separate host from port
+	host, _, err := net.SplitHostPort(p.Addr.String())
+	if err != nil {
+		return nil, false
+	}
+
+	// parse ip from host
+	sourceIP = net.ParseIP(host)
+	if sourceIP == nil {
+		return nil, false
+	}
+
+	return sourceIP, true
+}
+
+// PostSpans implements the jeager api_v2.CollectorServiceServer interface.  The grpc server will pass the jaeger
+// batches it receives to this method.  This method will convert the jaeger batches to SignalFx spans and pass them
+// on to the output writer.
+func (m *Monitor) PostSpans(ctx context.Context, r *api_v2.PostSpansRequest) (*api_v2.PostSpansResponse, error) {
+	if r != nil {
+		// convert the batch to SignalFx metrics
+		spans := jaegerprotobuf.JaegerProtoBatchToSFX(&r.Batch)
+
+		// tag the source on the span meta data
+		source, hasSource := extractRemoteAddressToContext(ctx)
+		if hasSource {
+			for i := range spans {
+				// monitor Output expects Meta to be non-nil
+				if spans[i].Meta == nil {
+					spans[i].Meta = map[interface{}]interface{}{}
+				}
+				spans[i].Meta[constants.DataSourceIPKey] = source
+			}
+		}
+
+		// send the spans on through the agent
+		m.Output.SendSpans(spans...)
+	}
+
+	return &api_v2.PostSpansResponse{}, nil
+}
+
+// Configure the monitor and kick off volume metric syncing
+func (m *Monitor) Configure(conf *Config) error {
+	// parse tls configurations
+	var grpcOptions []grpc.ServerOption
+	if conf.TLS != nil {
+		creds, err := conf.TLS.Credentials()
+		if err != nil {
+			return err
+		}
+		grpcOptions = []grpc.ServerOption{grpc.Creds(creds)}
+	}
+
+	// create the grpc server
+	m.grpc = grpc.NewServer(grpcOptions...)
+
+	var err error
+	// create a listener with the configured ListenAddress
+	m.ln, err = net.Listen("tcp", conf.ListenAddress)
```

Comment:
Move this into the goroutine and also implement a retry loop that retries indefinitely to listen on this port until shutdown.  The key thing is that if something happens to be listening on this port when the agent start but later stops listening, then the monitor should not just fail out at the beginning and then never work until an agent restart.  Make it log an error message whenever it can't listen since that is still a bad condition.

One of the core properties of monitors that we want is that they are persistent in the face of errors and can reach the desired monitoring state when it is possible to do so.